### PR TITLE
faster rcnn model for object detection

### PR DESCRIPTION
**Pull request by nikk-nikaznan**

### CODE REVIEW

File: crabs/detection_tracking/sort.py
Comment (sfmig) [inline crabs/detection_tracking/sort.py:1]: I was wondering if there is a more maintainable way of using SORT... I saw there is [a packaged version of SORT](https://github.com/kadirnar/sort-pip) (but there seem to be license issues?), but I think a nicer way would be to [git clone the SORT repo](https://github.com/abewley/sort/tree/master) right?

Alternatively we can check [deep SORT](https://github.com/nwojke/deep_sort) which seems like an improvement using appearance

File: crabs/detection_tracking/evaluate_model.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import torch
+from detection_utils import (
+    create_dataloader,
+    get_test_transform,
+    myFasterRCNNDataset,
+)
+from evaluate import test_detection
+
+# select device (whether GPU or CPU)
+device = (
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:14]: same comment about the scope as previously

File: crabs/detection_tracking/evaluate_model.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+)
+
+
+class Detector_Test:
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:19]: same comment about using test in classes and functions. I'm not sure if pytest would be confused in this case but I still find it confusing to have the word test in something that is not an actual test?

File: crabs/detection_tracking/evaluate_model.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    """
+    A class for testing object detection models using pre-trained classification.
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:21]:
```suggestion
    A class for running object detection models using pre-trained classification.
```
maybe?

File: crabs/detection_tracking/evaluate_model.py (hunk 2 above)
Comment (sfmig) [reply crabs/detection_tracking/evaluate_model.py:19]: maybe "model evaluation" or "detector evaluation"? I find I cannot tell very well what this class does from its name

File: crabs/detection_tracking/evaluate_model.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        Load the pre-trained subject classification model.
+        """
+        # Load the pre-trained subject predictor
+        # TODO: deal with different model
+        self.trained_model = torch.load(
+            self.args.model_dir, map_location=torch.device("cpu")
+        )
+
+    def _load_dataset(self) -> None:
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:64]: can we reuse the equivalent code from the detector utils? Maybe we can have a parent class that the evaluation class and the detector class inherit from? Right now it feels a bit duplicated

File: crabs/detection_tracking/evaluate_model.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        """Load images and annotation file for training"""
+
+        self.annotation = (
+            f"{self.main_dir}/annotations/VIA_JSON_combined_coco_gen.json"
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:68]: same as before, it would be nice not to have this hardcoded

File: crabs/detection_tracking/evaluate_model.py
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:1]: looks good, but it seems we could combine a lot with the train_model module? just to avoid code duplication, needing to edit things in two places or potentially bits getting out of sync

File: crabs/detection_tracking/models.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    # Get the number of input features
+    in_features = model.roi_heads.box_predictor.cls_score.in_features
+    # define a new head for the detector with required number of classes
+    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
+
+    return model
+
+
+def train_faster_rcnn(
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:48]: I feel that this could be better placed in train_model.py, what do you think?

File: crabs/detection_tracking/models.py
Comment (sfmig) [inline crabs/detection_tracking/models.py:1]: I like the idea of having a model factory kinda, maybe in the future we can add YOLO to the mix etc. Not sure this is the best way but I think it's a good starting point! We can ask Sam for software engineering advice

File: crabs/detection_tracking/models.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+
+def create_faster_rcnn(num_classes: int, coco_model: bool = True) -> nn.Module:
+    """Create faster rcnn model for the training
+
+    Parameters
+    ----------
+    num_classes : int
+        number of classes to train
+    coco_moder : bool
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:24]:
```suggestion
    coco_model : bool
```
I thought this meant whether the annotations will be in COCO format or not, but I think you mean whether to get the faster RCNN model with all the COCO classes as output, or with custom classes. Is that correct?

If so, maybe we could just have num_classes equal to None by default, and if it's None then we return the faster RCNN model with all the COCO classes.

File: crabs/detection_tracking/models.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        optimizer instance for the given model
+
+    Returns
+    -------
+    model : nn.Module
+        the trained model
+    """
+
+    # select device (whether GPU or CPU)
+    device = (
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:74]: nice! I like it here ✨ 😅

File: crabs/detection_tracking/models.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        torch.device("cuda")
+        if torch.cuda.is_available()
+        else torch.device("cpu")
+    )
+
+    # setup tensorboard stuff
+    layout = {
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:81]: nice, love being able to babysit the training with visualisation! 

I was wondering where is the MLFlow implementation? is it in a separate PR? (my understanding was that we were using MLflow to log so far, is that correct?)

File: crabs/detection_tracking/models.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    }
+    writer = SummaryWriter(f"/tmp/tensorboard/{int(time.time())}")
+    writer.add_custom_scalars(layout)
+
+    def log_scalar(name, value, epoch):
+        """Log a scalar value to both MLflow and TensorBoard"""
+        writer.add_scalar(name, value, epoch)
+        mlflow.log_metric(name, value)
+
+    EXPERIMENT_NAME = "baseline"
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:101]: should EXPERIMENT_NAME be an input to this function? I imagine we use it to identify different runs, so then hardcoding it here doesn't seem very useful right?

File: crabs/detection_tracking/models.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+    try:
+        EXPERIMENT_ID = mlflow.get_experiment_by_name(
+            EXPERIMENT_NAME
+        ).experiment_id
+    except mlflow.exception.MlflowException:
+        EXPERIMENT_ID = mlflow.create_experiment(EXPERIMENT_NAME)
+
+    with mlflow.start_run(experiment_id=EXPERIMENT_ID, run_name=RUN_NAME):
+        mlflow.log_params(config)
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:112]: oh nice! are we using both tensorboard and MLFlow? should we just use one?

File: crabs/detection_tracking/models.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+            for batch_idx, (imgs, annotations) in enumerate(train_dataloader):
+                i += 1
+                imgs = list(img.to(device) for img in imgs)
+                annotations = [
+                    {k: v.to(device) for k, v in t.items()}
+                    for t in annotations
+                ]
+
+                loss_dict = model(imgs, annotations)
+                losses = sum(loss for loss in loss_dict.values())
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:130]: I think it's slightly more standard to compute the mean loss of the batch right? (or why the sum?)

File: crabs/detection_tracking/models.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+                optimizer.zero_grad()
+                losses.backward()
+                optimizer.step()
+
+            log_scalar("total_loss/train", losses, epoch)
+
+        # Upload the TensorBoard event logs as a run artifact
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:138]: aaah I see! so the tensorboard logs can be seen in MLFlow dashboard? nice ✨

File: crabs/detection_tracking/inference_model.py [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+from pathlib import Path
+
+import cv2
+import numpy as np
+import torch
+import torchvision.transforms as transforms
+from sort import Sort
+
+# select device (whether GPU or CPU)
+device = (
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:12]: same bit about the scope as before 😬

File: crabs/detection_tracking/inference_model.py
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:1]: I think here there is also a bit of code duplication that we could avoid... maybe this is good stuff to do pair prog with Sam and get his wisdom!

File: crabs/detection_tracking/inference_model.py [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+    torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+)
+
+
+class Detector_Inference:
+    """
+    A class for performing object detection or tracking inference on a video
+    using a pre-trained model.
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:20]:
```suggestion
    using a trained model.
```
why pre-trained? 🤔

File: crabs/detection_tracking/inference_model.py [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+        _load_pretrain_model(self) -> None:
+            Load the pre-trained subject classification model.
+
+        __inference(self, frame, video_file, frame_id) -> None:
+            Perform inference on a single frame of the video.
+
+        _load_video(self) -> None:
+            Load the input video and perform inference on its frames.
+
+        inference_model(self) -> None:
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:43]:
```suggestion
        run_inference(self) -> None:
```
maybe more clear?

File: crabs/detection_tracking/inference_model.py [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+            Perform object detection or tracking inference on the input video.
+
+    """
+
+    def __init__(self, args: argparse.Namespace) -> None:
+        self.args = args
+        self.vid_dir = args.vid_dir
+        self.score_threshold = args.score_threshold
+        self.sort_crab = Sort()
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:52]: what is `sort_crab`? (can it be rename to something more self-explanatory?)

File: crabs/detection_tracking/inference_model.py [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+                prediction,
+                pred_score,
+                self.score_threshold,
+                self.sort_crab,
+            )
+        return frame_out
+
+    def _load_video(self) -> None:
+        """
+        Load the input video and perform inference on its frames.
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:222]: maybe also include in the docstring that the multi channels are extracted if required.

File: crabs/detection_tracking/inference_model.py [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+    """
+    A class for performing object detection or tracking inference on a video
+    using a pre-trained model.
+
+    Args:
+        args (argparse.Namespace): Command-line arguments containing
+        configuration settings.
+
+    Attributes:
+        args (argparse.Namespace): The command-line arguments provided.
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:27]: I think some of the args are missing right?

File: crabs/detection_tracking/inference_model.py [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+                max_abs_blurred_frame,
+            )
+        )
+
+        # compute the motion channel for frame_idx
+        return np.abs(
+            background_subtracted_frame_delta - background_subtracted_frame,
+        )
+
+    def __inference(
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:176]: maybe rename to
```suggestion
    def run_inference_one_frame(
```
?

File: crabs/detection_tracking/inference_model.py [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+            from _inference import inference_detection
+
+            frame_out = inference_detection(
+                frame, prediction, pred_score, self.score_threshold
+            )
+
+        else:
+            from _inference import inference_tracking
+
+            frame_out = inference_tracking(
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:211]: I don't understand very well how we can do tracking on one frame only 🤔

File: crabs/detection_tracking/inference_model.py [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+                frame,
+                prediction,
+                pred_score,
+                self.score_threshold,
+                self.sort_crab,
+            )
+        return frame_out
+
+    def _load_video(self) -> None:
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:220]:
```suggestion
    def run_inference_on_video(self) -> None:
```
what do you think?

File: crabs/detection_tracking/inference_model.py [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+                    mean_blurred_frame += blurred_frame
+
+                    # accumulate max absolute values
+                    max_abs_blurred_frame = np.maximum(
+                        max_abs_blurred_frame, abs(blurred_frame)
+                    )
+
+                    frame_count += 1
+            # compute the mean
+            mean_blurred_frame = mean_blurred_frame / frame_count
```
Comment (sfmig) [inline crabs/detection_tracking/inference_model.py:278]: can we reuse the utils for computing the additional channels here? To avoid code duplication and making sure we are using the same approach in training and inference.

If it's a lot for this PR feel free to open an issue

**Review by sfmig — APPROVED** [review]
Nice work! I did have quite a few comments (but also the PR is big 😅) 

I think going forward we should try to make smaller PRs (maybe only for the training work, then evaluation etc.) - I know it's hard tho!

I set it to "approve merging" but if you can make sure that any bits not addressed are made into issues I would really appreciate it 🙌 

Happy to chat about any of it offline if you need it!

File: crabs/detection_tracking/evaluate.py [hunk 24]
```diff
@@ -0,0 +1,273 @@
+from collections import defaultdict
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate.py]: We can, but some of the utils are shared between training and testing. Will add an issue for future me to think about

File: crabs/detection_tracking/evaluate_model.py
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate_model.py:1]: agree - especially on the data loader section - will add to the issue

File: crabs/detection_tracking/evaluate_model.py (hunk 2 above)
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate_model.py:19]: Yes, Sorry about that. the code used to be train and test. I changed some that failed the pytest. Will change them all to be consistent.
